app/dao/notifications_dao.py

```diff
@@ -114,10 +114,20 @@ def _update_notification_status(notification, status):
 @statsd(namespace="dao")
 @transactional
 def update_notification_status_by_id(
-        notification_id: uuid, status: str, sent_by: str = None, status_reason: str = None
+        notification_id: uuid,
+        status: str,
+        sent_by: str = None,
+        status_reason: str = None,
+        current_status: str = None
 ) -> Notification:
-    notification = Notification.query.with_for_update().filter(Notification.id == notification_id).first()
 
+    # the order of notification status that must be maintained
+    order_matrix = [NOTIFICATION_SENDING, NOTIFICATION_SENT, NOTIFICATION_DELIVERED]
+    notification = Notification.query.with_for_update().filter(Notification.id == notification_id)
+    if current_status is not None:
+        notification.filter(Notification.status == current_status)
```

Comment:
"notification" is a misleading variable name here.  At a glance, one might assume it is a Notification instance, but it's actually a query instance.  I would call it "notification_query".